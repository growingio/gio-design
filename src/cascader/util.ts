import { OptionProps } from '../list/interfance';

export const getLabelByValue = (val?: string | string[], opts?: OptionProps | OptionProps[], separator = '') =>
  val?.includes('.')
    ? (val as any)
        ?.split('.')
        ?.reduce(
          (prev: string[], curr: string) => [...prev, (opts as OptionProps[]).find((o) => o.value === curr)?.label],
          []
        )
        ?.join(separator)
    : (opts as OptionProps).label;

export default {
  getLabelByValue,
};
