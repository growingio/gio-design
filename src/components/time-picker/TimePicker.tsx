import React, { Component, ReactNode } from 'react';
import Trigger from 'rc-trigger';
import moment, { Moment } from 'moment';
import classNames from 'classnames';
import Panel from './Panel';
import placements, { defaultAlign } from './placements';
import { TimePickerProps } from './interface';

function noop() {/* ... */}

function refFn(field: string, component: ReactNode) {
  this[field] = component;
}

interface TimePickerState {
  open?: boolean;
  value?: Moment;
}

class Picker extends Component<TimePickerProps, TimePickerState> {
  public static defaultProps = {
    clearText: 'clear',
    prefixCls: 'rc-time-picker',
    defaultOpen: false,
    inputReadOnly: false,
    style: {},
    className: '',
    inputClassName: '',
    popupClassName: '',
    popupStyle: {},
    align: defaultAlign,
    defaultOpenValue: moment(),
    allowEmpty: true,
    showHour: true,
    showMinute: true,
    showSecond: true,
    disabledHours: noop,
    disabledMinutes: noop,
    disabledSeconds: noop,
    hideDisabledOptions: false,
    placement: 'bottomLeft',
    onChange: noop,
    onAmPmChange: noop,
    onOpen: noop,
    onClose: noop,
    onFocus: noop,
    onBlur: noop,
    addon: noop,
    use12Hours: false,
    focusOnOpen: false,
    onKeyDown: noop,
  };

  private saveInputRef: any;

  private savePanelRef: any;

  private picker: any;

  public constructor(props: TimePickerProps) {
    super(props);
    this.saveInputRef = refFn.bind(this, 'picker');
    this.savePanelRef = refFn.bind(this, 'panelInstance');
    const { defaultOpen, defaultValue, open = defaultOpen, value = defaultValue } = props;
    this.state = {
      open,
      value,
    };
  }

  public static getDerivedStateFromProps(props: TimePickerProps, state: TimePickerState) {
    const newState: any = {};
    if ('value' in props) {
      newState.value = props.value;
    }
    if (props.open !== undefined) {
      newState.open = props.open;
    }
    return Object.keys(newState).length > 0
      ? {
          ...state,
          ...newState,
        }
      : null;
  }

  private onPanelChange = (value: Moment) => {
    this.setValue(value);
  };

  private onAmPmChange = (ampm: any) => {
    const { onAmPmChange } = this.props;
    onAmPmChange(ampm);
  };

  private onClear = (event: any) => {
    event.stopPropagation();
    this.setValue(undefined);
    this.setOpen(false);
  };

  private onVisibleChange = (open: boolean) => {
    this.setOpen(open);
  };

  private onEsc = () => {
    this.setOpen(false);
    this.focus();
  };

  private onKeyDown = (e: any) => {
    if (e.keyCode === 40) {
      this.setOpen(true);
    }
  };

  private setValue(value?: Moment) {
    const { onChange } = this.props;
    if (!('value' in this.props)) {
      this.setState({
        value,
      });
    }
    if (onChange) {
      onChange(value);
    }
  }

  private getFormat() {
    const { format, showHour, showMinute, showSecond, use12Hours } = this.props;
    if (format) {
      return format;
    }

    if (use12Hours) {
      const fmtString = [showHour ? 'h' : '', showMinute ? 'mm' : '', showSecond ? 'ss' : '']
        .filter((item) => !!item)
        .join(':');

      return fmtString.concat(' a');
    }

    return [showHour ? 'HH' : '', showMinute ? 'mm' : '', showSecond ? 'ss' : ''].filter((item) => !!item).join(':');
  }

  private getPanelElement() {
    const {
      prefixCls,
      placeholder,
      disabledHours,
      disabledMinutes,
      disabledSeconds,
      hideDisabledOptions,
      inputReadOnly,
      showHour,
      showMinute,
      showSecond,
      defaultOpenValue,
      clearText,
      addon,
      use12Hours,
      focusOnOpen,
      onKeyDown,
      hourStep,
      minuteStep,
      secondStep,
      clearIcon,
    } = this.props;
    const { value } = this.state;
    return (
      <Panel
        clearText={clearText}
        prefixCls={`${prefixCls}-panel`}
        ref={this.savePanelRef}
        value={value}
        inputReadOnly={inputReadOnly}
        onChange={this.onPanelChange}
        onAmPmChange={this.onAmPmChange}
        defaultOpenValue={defaultOpenValue}
        showHour={showHour}
        showMinute={showMinute}
        showSecond={showSecond}
        onEsc={this.onEsc}
        format={this.getFormat()}
        placeholder={placeholder}
        disabledHours={disabledHours}
        disabledMinutes={disabledMinutes}
        disabledSeconds={disabledSeconds}
        hideDisabledOptions={hideDisabledOptions}
        use12Hours={use12Hours}
        hourStep={hourStep}
        minuteStep={minuteStep}
        secondStep={secondStep}
        addon={addon}
        focusOnOpen={focusOnOpen}
        onKeyDown={onKeyDown}
        clearIcon={clearIcon}
      />
    );
  }

  private getPopupClassName() {
    const { showHour, showMinute, showSecond, use12Hours, prefixCls, popupClassName } = this.props;
    let selectColumnCount = 0;
    if (showHour) {
      selectColumnCount += 1;
    }
    if (showMinute) {
      selectColumnCount += 1;
    }
    if (showSecond) {
      selectColumnCount += 1;
    }
    if (use12Hours) {
      selectColumnCount += 1;
    }
    // Keep it for old compatibility
    return classNames(
      popupClassName,
      {
        [`${prefixCls}-panel-narrow`]: (!showHour || !showMinute || !showSecond) && !use12Hours,
      },
      `${prefixCls}-panel-column-${selectColumnCount}`
    );
  }

  private setOpen(open: boolean) {
    const { onOpen, onClose } = this.props;
    const { open: currentOpen } = this.state;
    if (currentOpen !== open) {
      if (!('open' in this.props)) {
        this.setState({ open });
      }
      if (open) {
        onOpen && onOpen({ open });
      } else {
        onClose && onClose({ open });
      }
    }
  }

  private focus() {
    this.picker.focus();
  }

  // private blur() {
  //   this.picker.blur();
  // }

  private renderClearButton() {
    const { value } = this.state;
    const { prefixCls, allowEmpty, clearIcon, clearText, disabled } = this.props;
    if (!allowEmpty || !value || disabled) {
      return null;
    }

    if (React.isValidElement(clearIcon)) {
      const { onClick } = clearIcon.props || {};
      const clearIconNode = React.cloneElement(clearIcon, {
        onClick: (...args: any) => {
          if (onClick) {
            onClick(...args);
          }
          if (args && args[0]) {
            this.onClear(args[0]);
          }
        },
      });

      return (
        <span role="button" className={`${prefixCls}-clear`}>
          {clearIconNode}
        </span>
      );
    }

    return (
      <button type="button" className={`${prefixCls}-clear`} title={clearText} onClick={this.onClear} tabIndex={0} aria-hidden="true">
        {clearIcon}
      </button>
    );
  }

  public render() {
    const {
      prefixCls,
      placeholder,
      placement = 'bottomRight',
      id,
      disabled,
      transitionName,
      className,
      inputReadOnly,
      name,
      inputIcon,
      inputClassName,
      getPopupContainer,
      autoComplete,
      onFocus,
      onBlur,
      autoFocus,
      style,
      align,
      popupStyle,
    } = this.props;
    const { open, value } = this.state;
    const popupClassName = this.getPopupClassName();
    return (
      <Trigger
        prefixCls={`${prefixCls}-panel`}
        popupClassName={popupClassName}
        popupStyle={popupStyle}
        popup={this.getPanelElement()}
        popupAlign={align}
        builtinPlacements={placements}
        popupPlacement={placement}
        action={disabled ? [] : ['click']}
        destroyPopupOnHide
        getPopupContainer={getPopupContainer}
        popupTransitionName={transitionName}
        popupVisible={open}
        onPopupVisibleChange={this.onVisibleChange}
      >
        <span className={classNames(prefixCls, className)} style={style}>
          <input
            className={classNames(`${prefixCls}-input`, inputClassName, {
              [`${prefixCls}-input-focus`]: open,
            })}
            ref={this.saveInputRef}
            type="text"
            placeholder={placeholder}
            name={name}
            onKeyDown={this.onKeyDown}
            disabled={disabled}
            value={(value && value.format(this.getFormat())) || ''}
            autoComplete={autoComplete}
            onFocus={onFocus}
            onBlur={onBlur}
            autoFocus={autoFocus} // eslint-disable-line jsx-a11y/no-autofocus
            onChange={noop}
            readOnly={!!inputReadOnly}
            id={id}
            required
          />
          <span role="button" className={`${prefixCls}-addon-icon`}>
            {inputIcon}
          </span>
          {this.renderClearButton()}
        </span>
      </Trigger>
    );
  }
}

export default Picker;
