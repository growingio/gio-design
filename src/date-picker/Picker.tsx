import React from 'react';
import classnames from 'classnames';
import { usePrefixCls, useControlledState } from '@gio-design/utils';
import { format } from 'date-fns/fp';
import { CalendarOutlined } from '@gio-design/icons';
import Popover from '../popover';
import { InputButton } from '../input';
import StaticDatePicker from '../static-date-picker';
import { DatePickerProps } from './interfaces';

export const DatePicker: React.FC<DatePickerProps> = (props: DatePickerProps) => {
  const {
    onVisibleChange: onDropdownVisibleChange,
    overlayClassName,
    visible: dropdownVisible,
    trigger,
    disabled,
    value,
    defaultValue,
    onSelect,
    disabledDate,
    placeholder,
    allowClear,
    format: formatString,
    ...restProps
  } = props;

  const prefixCls = usePrefixCls('date-picker-new');
  const overlayCls = classnames(`${prefixCls}-overlay`, overlayClassName);

  const [visible, setVisible] = useControlledState(dropdownVisible, false);

  const [controlledValue, setControlledValue] = useControlledState(value, defaultValue);

  const formatDate = (date: Date) => format(formatString ?? 'yyyy/MM/dd', date);

  const handleVisibleChange = (current: boolean) => {
    setVisible(current);
    onDropdownVisibleChange?.(current);
  };

  const handleOnSelect = (currentValue: Date) => {
    setControlledValue(currentValue);
    setVisible(false);
    onSelect?.(currentValue, formatDate(currentValue));
  };

  const content = <StaticDatePicker onSelect={handleOnSelect} disabledDate={disabledDate} />;

  function renderTrigger() {
    if (trigger) {
      return <div>{trigger}</div>;
    }
    return (
      <InputButton
        prefix={<CalendarOutlined />}
        placeholder={placeholder}
        disabled={disabled}
        allowClear={allowClear}
        value={controlledValue && formatDate(controlledValue)}
      />
    );
  }

  return (
    <Popover
      content={content}
      trigger={['click', 'focus']}
      visible={visible}
      placement="bottomLeft"
      overlayClassName={overlayCls}
      onVisibleChange={handleVisibleChange}
      {...restProps}
    >
      {renderTrigger()}
    </Popover>
  );
};

export default DatePicker;
