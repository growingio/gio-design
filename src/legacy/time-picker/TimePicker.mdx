import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Select from './Select';

# Time Picker 时分选择器

<Subtitle>当用户需要输入一个时间，可以点击触发选框，弹出时间面板进行选择。</Subtitle>

## 参数说明

<ArgsTable of={Select} />

## 代码演示

### 默认

[Example](/?path=/story/functional-components-timepicker--default)

### 时分秒

[Example](/?path=/story/functional-components-timepicker--seconds)
