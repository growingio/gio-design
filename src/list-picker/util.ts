import { OptionProps, SelectionItemProps } from '../list/interfance';

export const getFlattenOptions = (
  data: OptionProps[] | SelectionItemProps[],
  isSelection = false
): { groupName: string; groupId: string; options: OptionProps[] }[] | OptionProps[] => {
  const groupMap = new Map();
  if (!isSelection) {
    return data;
  }
  (data as SelectionItemProps[])?.map((cur: SelectionItemProps) => {
    const gValue = groupMap.get(cur.groupId);
    if (gValue) {
      const { options, ...rest } = gValue;
      return groupMap.set(cur.groupId, {
        options: [...options, cur],
        ...rest,
      });
    }
    return groupMap.set(cur.groupId, {
      label: cur.groupName,
      value: cur.groupId,
      options: [cur],
    });
  });
  const flattenOption: { groupName: string; groupId: string; options: OptionProps[] }[] = [];
  groupMap.forEach((value) => {
    flattenOption.push({
      groupName: value.label,
      groupId: value.value,
      options: value.options ?? [],
    });
  });
  return flattenOption;
};

export default {
  getFlattenOptions,
};
