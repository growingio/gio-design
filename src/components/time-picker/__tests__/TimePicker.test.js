import React from 'react';
import moment, { Moment } from 'moment';
import { mount, render } from 'enzyme';
import { noop } from 'lodash';
import ReactDOM from 'react-dom';
import Picker from '../index';
import TimePicker from '../TimePicker';
import { clickInput, clickSelectItem, matchValue } from './util';

// 打印快照
describe('Testing timepicker', () => {
  it('should match alert base snapshot.', () => {
    // const wrapper = render(<Picker showMinute />);
    // expect(wrapper).toMatchSnapshot();
  });
  // 测试组件是否正常渲染
  it('should render a DOM', () => {
    const wrapper = mount(<Picker className="test-cls" />);
    expect(wrapper.find('.gio-time-picker')).toHaveLength(1);
  });
  // 测试点击输入框时是否弹面板框
  it('panel should popup corretly', () => {
    const wrapper = mount(<Picker />);
    wrapper.find('.gio-time-picker-input').simulate('click');
    expect(wrapper.exists('.gio-time-picker-panel')).toBe(true);
  });
  // 挂载属性，卸载不报错
  it('should be mount, setProps, unmount with no error', () => {
    expect(() => {
      const wrapper = mount(<Picker />);
      wrapper.setProps({ inputClassName: 'inputclass' });
      wrapper.setProps({ use12Hours: true });
      wrapper.unmount();
    }).not.toThrow();
  });
  //  props test
  it('props test', async () => {
    const wrapper = mount(<Picker />);
    await wrapper.find('.gio-time-picker-input').simulate('click');
    const onClickMock = jest.fn();

    expect(
      mount(
        <Picker
          format="HH:mm:ss"
          placeholder="请选择时间"
          onChange={onClickMock}
          clearText="clear"
          defaultOpen={false}
          inputReadOnly={false}
          //   style={margin:auto}
          className=""
          inputClassName=""
          popupClassName=""
          //   popupStyle={}
          allowEmpty
          showHour
          showMinute
          showSecond
          hideDisabledOptions={false}
          placement="bottomLeft"
          use12Hours={false}
          focusOnOpen={false}
          //   defaultOpen={false}
          value={Moment}
          onAmPmChange={onClickMock}
          onPanelChange={onClickMock}
          onClear={onClickMock}
          onEsc={onClickMock}
          onKeyDown={onClickMock}
        />
      ).find('.gio-time-picker-panel-select')
    ).toMatchSnapshot();
  });
  // 测试选择面板数据之前是否为空
  it('before select value is none ', () => {
    const wrapper = mount(<Picker />);
    expect(wrapper.find('.gio-time-picker-input').text()).toBe('');
  });
  // 测试点击输入框后，弹出面板数据是否正确
  it('triggle and value is incurrent ', () => {
    const wrapper = mount(<Picker />);

    wrapper.find('.gio-time-picker-input').simulate('click');
    const lis = wrapper.find('.gio-time-picker-panel-select');
    expect(lis.at(0).find('li').length).toBe(24);
    expect(lis.at(1).find('li').length).toBe(60);
  });

  it('timepciker match snapshot1', () => {
    expect(
      <TimePicker
        format="HH:mm"
        placeholder="请选择时间"
        showSecond={false}
        onChange={noop}
        value={moment('2011-03-05T00:00:00.000Z')}
        defaultOpenValue={null}
      />
    ).toMatchSnapshot()
  })

  it('timepciker match snapshot2', () => {
    expect(
      <TimePicker
        format="HH:mm:ss"
        placeholder="请选择时间"
        showSecond
        onChange={noop}
        value={moment('2011-03-05T00:00:00.000Z')}
        defaultOpenValue={null}
      />
    ).toMatchSnapshot()
  })
});

describe('TimePicker', () => {
  let container;

  function renderPicker(props, options) {
    const showSecond = true;
    const format = 'HH:mm:ss';

    return mount(
      <TimePicker
        format={format}
        showSecond={showSecond}
        defaultValue={moment('12:57:58', format)}
        {...props}
      />,
      options,
    );
  }

  function renderPickerWithoutSeconds(props) {
    const showSecond = false;
    const format = 'HH:mm';

    return mount(
      <TimePicker
        format={format}
        showSecond={showSecond}
        defaultValue={moment('08:24', format)}
        {...props}
      />,
    );
  }

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(container);
    document.body.removeChild(container);
  });

  describe('render panel to body', () => {
    it('popup correctly', async () => {
      const onChange = jest.fn();
      const picker = renderPicker({
        onChange,
      });
      expect(picker.state().open).toBeFalsy();
      matchValue(picker, '12:57:58');
      clickInput(picker);

      expect(picker.state().open).toBeTruthy();
      clickSelectItem(picker, 0, 1);

      expect(onChange).toBeCalled();
      expect(onChange.mock.calls[0][0].hour()).toBe(1);
      expect(onChange.mock.calls[0][0].minute()).toBe(57);
      expect(onChange.mock.calls[0][0].second()).toBe(58);
      matchValue(picker, '01:57:58');
      expect(picker.state().open).toBeTruthy();
    });

    it('destroy correctly', async () => {
      const picker = renderPicker({}, { attachTo: container });
      expect(picker.state().open).toBeFalsy();
      clickInput(picker);
      expect(picker.state().open).toBeTruthy();

      expect(document.querySelectorAll('.gio-time-picker').length).not.toBe(0);
      expect(picker.find('Panel li').length).toBeTruthy();
      picker.detach();

      expect(document.querySelectorAll('.gio-time-picker').length).toBe(0);
      expect(picker.instance().panelInstance).toBeFalsy();
    });

    it('support name', () => {
      const picker = renderPicker({
        name: 'time-picker-form-name',
      });
      expect(picker.find('.gio-time-picker-input').instance().name).toBe('time-picker-form-name');
    });

    it('support focus', () => {
      const picker = renderPicker({
        name: 'time-picker-form-name',
      });
      expect(typeof picker.instance().focus).toBe('function');
    });

    it('should be controlled by open', () => {
      const picker = renderPicker({
        open: false,
      });
      expect(picker.state().open).toBeFalsy();
      clickInput(picker);
      expect(picker.state().open).toBeFalsy();
    });

    it('support custom icon', () => {
      const picker = renderPicker({
        inputIcon: 'test-select',
      });
      expect(picker.find('.gio-time-picker').text()).toBe('test-select');
    });
  });

  describe('render panel to body (without seconds)', () => {
    it('popup correctly', async () => {
      const onChange = jest.fn();
      const picker = renderPickerWithoutSeconds({
        onChange,
      });
      expect(picker.state().open).toBeFalsy();
      matchValue(picker, '08:24');
      clickInput(picker);

      expect(picker.find('.gio-time-picker-panel-inner').length).toBeTruthy();
      expect(picker.state().open).toBeTruthy();
      clickSelectItem(picker, 0, 1);

      expect(onChange).toBeCalled();
      expect(onChange.mock.calls[0][0].hour()).toBe(1);
      expect(onChange.mock.calls[0][0].minute()).toBe(24);
      matchValue(picker, '01:24');
      expect(picker.state().open).toBeTruthy();
    });
  });

  describe('render panel to body 12pm mode', () => {
    it('popup correctly', async () => {
      const onChange = jest.fn();
      const picker = renderPickerWithoutSeconds({
        use12Hours: true,
        value: null,
        onChange,
      });
      expect(picker.state().open).toBeFalsy();
      matchValue(picker, '');
      clickInput(picker);

      expect(picker.find('.gio-time-picker-panel-inner').length).toBeTruthy();
      expect(picker.state().open).toBeTruthy();
      clickSelectItem(picker, 0, 1);

      expect(onChange).toBeCalled();
      expect(picker.state().open).toBeTruthy();
    });
  });

  describe('other operations', () => {
    it('focus/blur correctly', async () => {
      let focus = false;
      let blur = false;

      const picker = renderPicker({
        onFocus: () => {
          focus = true;
        },
        onBlur: () => {
          blur = true;
        },
      });
      expect(picker.state().open).toBeFalsy();
      picker.find('.gio-time-picker-input').simulate('focus');
      expect(picker.state().open).toBeFalsy();
      picker.find('.gio-time-picker-input').simulate('blur');

      expect(focus).toBeTruthy();
      expect(blur).toBeTruthy();
    });
  });

  describe('allowEmpty', () => {
    it('should allow clear', async () => {
      const picker = renderPicker({
        allowEmpty: true,
      });
      expect(picker.render()).toMatchSnapshot();
    });

    it('cannot allow clear when disabled', async () => {
      const picker = renderPicker({
        allowEmpty: true,
        disabled: true,
      });
      expect(picker.render()).toMatchSnapshot();
    });
  });
});