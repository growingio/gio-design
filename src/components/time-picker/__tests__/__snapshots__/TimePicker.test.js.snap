// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing timepicker props test 1`] = `ReactWrapper {}`;

exports[`Testing timepicker timepciker match snapshot1 1`] = `
<Picker
  addon={[Function]}
  align={
    Object {
      "offset": Array [
        0,
        4,
      ],
      "overflow": Object {
        "adjustX": 1,
        "adjustY": 1,
      },
    }
  }
  allowEmpty={true}
  className=""
  clearText="clear"
  defaultOpen={false}
  defaultOpenValue={null}
  disabledHours={[Function]}
  disabledMinutes={[Function]}
  disabledSeconds={[Function]}
  focusOnOpen={false}
  format="HH:mm"
  hideDisabledOptions={false}
  inputClassName=""
  inputReadOnly={false}
  onAmPmChange={[Function]}
  onBlur={[Function]}
  onChange={[Function]}
  onClose={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onOpen={[Function]}
  placeholder="请选择时间"
  placement="bottomLeft"
  popupClassName=""
  popupStyle={Object {}}
  prefixCls="gio-time-picker"
  showHour={true}
  showMinute={true}
  showSecond={false}
  style={Object {}}
  use12Hours={false}
  value={"2011-03-05T00:00:00.000Z"}
/>
`;

exports[`Testing timepicker timepciker match snapshot2 1`] = `
<Picker
  addon={[Function]}
  align={
    Object {
      "offset": Array [
        0,
        4,
      ],
      "overflow": Object {
        "adjustX": 1,
        "adjustY": 1,
      },
    }
  }
  allowEmpty={true}
  className=""
  clearText="clear"
  defaultOpen={false}
  defaultOpenValue={null}
  disabledHours={[Function]}
  disabledMinutes={[Function]}
  disabledSeconds={[Function]}
  focusOnOpen={false}
  format="HH:mm:ss"
  hideDisabledOptions={false}
  inputClassName=""
  inputReadOnly={false}
  onAmPmChange={[Function]}
  onBlur={[Function]}
  onChange={[Function]}
  onClose={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onOpen={[Function]}
  placeholder="请选择时间"
  placement="bottomLeft"
  popupClassName=""
  popupStyle={Object {}}
  prefixCls="gio-time-picker"
  showHour={true}
  showMinute={true}
  showSecond={true}
  style={Object {}}
  use12Hours={false}
  value={"2011-03-05T00:00:00.000Z"}
/>
`;

exports[`TimePicker allowEmpty cannot allow clear when disabled 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "gio-time-picker",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "gio-time-picker-input",
          "disabled": "",
          "required": "",
          "type": "text",
          "value": "12:57:58",
        },
        "children": Array [],
        "name": "input",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-time-picker-addon-icon",
            "role": "button",
          },
          "children": Array [],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "role": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "role": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "disabled": undefined,
          "required": undefined,
          "type": undefined,
          "value": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "disabled": undefined,
          "required": undefined,
          "type": undefined,
          "value": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-time-picker-addon-icon",
          "role": "button",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-time-picker-input",
            "disabled": "",
            "required": "",
            "type": "text",
            "value": "12:57:58",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "disabled": undefined,
            "required": undefined,
            "type": undefined,
            "value": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "disabled": undefined,
            "required": undefined,
            "type": undefined,
            "value": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "role": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "role": undefined,
        },
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`TimePicker allowEmpty should allow clear 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "gio-time-picker",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "gio-time-picker-input",
          "required": "",
          "type": "text",
          "value": "12:57:58",
        },
        "children": Array [],
        "name": "input",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-time-picker-addon-icon",
            "role": "button",
          },
          "children": Array [],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "aria-hidden": "true",
              "class": "gio-time-picker-clear",
              "tabindex": "0",
              "title": "clear",
              "type": "button",
            },
            "children": Array [],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "aria-hidden": undefined,
              "class": undefined,
              "tabindex": undefined,
              "title": undefined,
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "aria-hidden": undefined,
              "class": undefined,
              "tabindex": undefined,
              "title": undefined,
              "type": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "role": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "role": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "required": undefined,
          "type": undefined,
          "value": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "required": undefined,
          "type": undefined,
          "value": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-time-picker-addon-icon",
          "role": "button",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "aria-hidden": "true",
            "class": "gio-time-picker-clear",
            "tabindex": "0",
            "title": "clear",
            "type": "button",
          },
          "children": Array [],
          "name": "button",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "aria-hidden": undefined,
            "class": undefined,
            "tabindex": undefined,
            "title": undefined,
            "type": undefined,
          },
          "x-attribsPrefix": Object {
            "aria-hidden": undefined,
            "class": undefined,
            "tabindex": undefined,
            "title": undefined,
            "type": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-time-picker-input",
            "required": "",
            "type": "text",
            "value": "12:57:58",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "required": undefined,
            "type": undefined,
            "value": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "required": undefined,
            "type": undefined,
            "value": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "role": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "role": undefined,
        },
      },
      Object {
        "attribs": Object {
          "aria-hidden": "true",
          "class": "gio-time-picker-clear",
          "tabindex": "0",
          "title": "clear",
          "type": "button",
        },
        "children": Array [],
        "name": "button",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-time-picker-addon-icon",
            "role": "button",
          },
          "children": Array [],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "gio-time-picker-input",
              "required": "",
              "type": "text",
              "value": "12:57:58",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "required": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "required": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "role": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "role": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "tabindex": undefined,
          "title": undefined,
          "type": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "tabindex": undefined,
          "title": undefined,
          "type": undefined,
        },
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
